File: src/main/java/work/nekow/particledrawing/api/TransformOp.java
package work.nekow.particledrawing.api;

import net.minecraft.world.phys.Vec3;

/**
 * Describes a transform operation applied to a {@link ParticleGroup}.
 * Immutable, designed for network serialization.
 */
public final class TransformOp {

    private final Type type;
    private final Vec3 delta;
    private final Vec3 axis;
    private final double radians;
    private final Vec3 pivot;
    private final Color targetColor;
    private final float targetScale;

    private TransformOp(Type type, Vec3 delta, Vec3 axis, double radians,
                        Vec3 pivot, Color targetColor, float targetScale) {
        this.type = type;
        this.delta = delta;
        this.axis = axis;
        this.radians = radians;
        this.pivot = pivot;
        this.targetColor = targetColor;
        this.targetScale = targetScale;
    }

    public Type type() { return type; }
    public Vec3 delta() { return delta; }
    public Vec3 axis() { return axis; }
    public double radians() { return radians; }
    public Vec3 pivot() { return pivot; }
    public Color targetColor() { return targetColor; }
    public float targetScale() { return targetScale; }

    public static TransformOp translate(Vec3 delta, Vec3 pivot) {
        return new TransformOp(Type.TRANSLATE, delta, null, 0, pivot, null, 0);
    }

    public static TransformOp rotate(Vec3 axis, double radians, Vec3 pivot) {
        return new TransformOp(Type.ROTATE, null, axis.normalize(), radians, pivot, null, 0);
    }

    public static TransformOp recolor(Color targetColor) {
        return new TransformOp(Type.RECOLOR, null, null, 0, null, targetColor, 0);
    }

    public static TransformOp scale(float targetScale, Vec3 pivot) {
        return new TransformOp(Type.SCALE, null, null, 0, pivot, null, targetScale);
    }

    public enum Type {
        TRANSLATE,
        ROTATE,
        RECOLOR,
        SCALE
    }

    @Override
    public String toString() {
        return "TransformOp{" + type + "}";
    }
}

File: src/main/kotlin/work/nekow/particledrawing/api/TransformOp.kt
package work.nekow.particledrawing.api

import net.minecraft.world.phys.Vec3

/**
 * 描述应用于 [ParticleGroup] 的变换操作。
 * 不可变，为网络序列化而设计。
 */
@Suppress("unused")
class TransformOp private constructor(
    val type: Type,
    val delta: Vec3?,
    val axis: Vec3?,
    val radians: Double,
    val pivot: Vec3?,
    val targetColor: Color?,
    val targetScale: Float
) {
    enum class Type {
        /** 平移 */
        TRANSLATE,
        /** 绕轴旋转 */
        ROTATE,
        /** 重着色 */
        RECOLOR,
        /** 缩放 */
        SCALE
    }

    override fun toString() = "TransformOp{$type}"

    companion object {
        fun translate(delta: Vec3, pivot: Vec3?): TransformOp {
            return TransformOp(Type.TRANSLATE, delta, null, 0.0, pivot, null, 0f)
        }

        fun rotate(axis: Vec3, radians: Double, pivot: Vec3?): TransformOp {
            return TransformOp(Type.ROTATE, null, axis.normalize(), radians, pivot, null, 0f)
        }

        fun recolor(targetColor: Color): TransformOp {
            return TransformOp(Type.RECOLOR, null, null, 0.0, null, targetColor, 0f)
        }

        fun scale(targetScale: Float, pivot: Vec3?): TransformOp {
            return TransformOp(Type.SCALE, null, null, 0.0, pivot, null, targetScale)
        }
    }
}
